Halogen

cryogenian opened the pull request:
I think it resolves #1, #4, #8. 

Please make another ticket with __move/rename__ and __share__ buttons.

jdegoes commented:
Can you squash down all these commits?

Review by @garyb (organization, names, types, etc) and @paf31 (Halogen usage, organization, types, etc.).

jdegoes commented:
@cryogenian Looks like there are conflicts, can you resolve them? Thanks!

cryogenian commented:
Oops, resolved

paf31 commented on src/Api/Fs.purs:
```diff
@@ -55,84 +34,108 @@ metadata path callback = do
+            Eff (dom :: DOM|e) Unit 
+metadata path callback = 
+  metadataF path $ \f -> do
+    let readed = readProp "children" f >>= readArray >>= traverse M.readItem
```
Maybe define an `IsForeign` instance for `M.Item`?

cryogenian replied:
`M.Item` is closed record. I know that using `newtype` cost nothing in runtime. But since it's closed and using record can reduce code size (i.e. not pattern match everytime `(Item item)`) and `M.Item` isn't used as part of larger structure, I've made `readItem` function.

paf31 commented on src/Controller/Driver.purs:
```diff
@@ -0,0 +1,233 @@
+outside driver = handleRoute driver 
+
+-- | Routing schema
+data Routes = SortAndQ M.Sort S.SearchQuery | SortOnly M.Sort | Index 
```
Without fully understanding the difference yet, maybe `Sort M.Sort (Maybe S.SearchQuery)` would make sense here?

paf31 commented on src/Model.purs:
```diff
@@ -1,58 +1,75 @@
+-- | Input messages 
+data Input
+  = Sorting Sort
+  | BreadcrumbUpdate [Breadcrumb]
```
Are the breadcrumbs not a function of the `path`?

paf31 replied:
I was thinking about trying to write a library using some kind of zipper comonad to deal with breadcrumbs and site navigation actually. I think that would fit well with the way the state works in Halogen.

cryogenian replied:
Sounds great!

paf31 commented on src/View.purs:
```diff
@@ -1,113 +1,202 @@
+import qualified Data.StrMap as SM
+import qualified Config as Config
+
+link :: forall a i node. (H.HTMLRepr node) => [A.Attr i] -> [node a i] -> node a i
```
Maybe consider using `Halogen.HTML.Target.target`?

paf31 commented:
Generally, I think the Halogen code looks good. I see that you're using a pre-`Component` commit of `purescript-halogen`, but I wonder if it would be possible to break things up into components once we make a release and this gets updated to use latest. Specifically, you have a fairly large sum type of inputs, and quite a few parts which have the same input and request types, and I can only imagine those types will grow over time. I'll take a look and see if there are any obvious places to break things up.

paf31 commented on src/Controller/Request.purs:
```diff
@@ -0,0 +1,173 @@
+
+handler :: forall e. M.Request ->
+           Aff.Aff (Hl.HalogenEffects (timer::Tm.Timer|e)) M.Input
+handler r = Aff.makeAff $ \_ k -> do 
```
Using `makeAff` at the top level here seems to be somewhat defeating the purpose of using `Aff`. Maybe `Aff` needs `callCC` to support the cases where we are providing multiple inputs at once, but it also seems like it might be possible to cut down on the number of inputs being used.

cryogenian replied:
handler` function will look better after switching to `affjax` I hope

jdegoes replied:
Let's chat about `callCC` when you get a chance. I have some reservations about it (but still open to adding it if all works out).

paf31 commented on src/Controller/Request.purs:
```diff
@@ -0,0 +1,173 @@
+      U.log "configure"
+
+    -- triggered when files changed in file selecting dialog
+    M.FileListChanged node state -> do
```
It seems like a bit of a code smell to have a `Node` in your input type, but there doesn't seem to be a much better option right now. 

<thinking out loud>

Maybe this is something that we can try to support in a similar way to how `EventHandler` supports things like `preventDefault`. Maybe some sort of extensible sum of actions would work, so that things like `clearValue` can be modeled in pure code.

cryogenian replied:
It would be great not to have external dependencies for `DOM` handling.

cryogenian commented:
* Changed `javascript:void(0)` to `target`
* removed `Breadcrumbs` message
* renamed `SortOnly` to `Sort`

jdegoes commented on src/Api/Fs.purs:
```diff
@@ -55,84 +34,108 @@ metadata path callback = do
+    either (const $ pure unit) callback readed 
+          
+          
+foreign import f2jsonImpl """
```
I'd prefer we use Argonaut here rather than rely on FFI "magic", as it's much more type safe and only a bit more verbose.

jdegoes commented on src/Api/Fs.purs:
```diff
@@ -55,84 +34,108 @@ metadata path callback = do
-            
 
+-- | gets current directory children
+metadata :: forall e. String -> ([M.Item] -> Eff (dom :: DOM | e) Unit) ->
```
Callbacks can be replaced with `Aff`. In fact, now with the (beta) release of Affjax, you may be able to simplify a lot of this code.

jdegoes commented on src/Api/Fs.purs:
```diff
@@ -55,84 +34,108 @@ metadata path callback = do
 
 
-post :: forall e. String -> Json -> ([Json] -> Eff (dom::DOM) Unit) ->
+move :: forall e. String -> String -> (Boolean -> Eff (dom::DOM|e) Unit) ->
```
We should consider introducing a `Path` newtype if it doesn't already exist somewhere.

garyb replied:
There's a `FilePath` type synonym in `purescript-node-path` but nothing suitable for this that I know of in the core/contrib libraries.

jdegoes commented on src/Config.purs:
```diff
@@ -16,6 +19,11 @@ searchTimeout = 500
 userEnabled = false
+
+newNotebookName :: String
+newNotebookName = "NewNotebook.nb"
```
I suggest new notebooks be titled "Untitled Notebook.slam" (similar to the convention of Google Docs). We may want to hide extensions so the user can't change them (at least hide for editing).

jdegoes commented on src/Config.purs:
```diff
@@ -16,6 +19,11 @@ searchTimeout = 500
+newNotebookName = "NewNotebook.nb"
+
+newFolderName :: String
+newFolderName = "NewFolder"
```
Similarly here, "Untitled Folder". When a user creates a new resource, ideally we'll implement inline editing so they can choose the name right away (or maybe use the Move / Rename dialog with a different header to choose the name).

jdegoes commented on src/Api/Fs.purs:
```diff
@@ -1,53 +1,32 @@
 
-metadata :: forall e. String -> ([Metadata] -> Eff (dom::DOM|e) Unit) -> 
+-- gets raw metadata
+metadataF :: forall e. String -> (Foreign -> Eff (dom::DOM|e) Unit) -> 
```
Nasty callback alert! Hopefully this whole module gets ported over to `Aff`.

jdegoes commented on src/Controller/Driver.purs:
```diff
@@ -0,0 +1,227 @@
+          driver $ M.ItemsUpdate items
+          driver $ M.Sorting sort
+          
+  where cleanPath :: String -> String
```
This is another reason we need a `Path` abstraction / library, so it can always keep them in canonical form.

jdegoes commented on src/Controller/Driver.purs:
```diff
@@ -0,0 +1,227 @@
+    -- since we use _minimatch_ to check `Contains` predicate
+    -- `Like` predicate works exactly like `Contains` if we
+    -- replace `% -> *` and `_ -> ?`
+    S.Like s -> flip check prj $ S.Contains $ S.Text $ like2glob s
```
We'll need to re-evaluate these semantics later after the search cell implementation to ensure consistency, but for now looks OK to me.

jdegoes commented on src/Controller/Driver.purs:
```diff
@@ -0,0 +1,227 @@
+-- | Set _sort_ in route
+data SortSetter = SortSetter M.Sort
+
+instance routeModifierSortSetter :: RS.RouteModifier SortSetter where
```
I know this is really an issue for `purescript-router`, but does `RouteModifier` have to be a type class? Is there any benefit in that approach that I don't see???

jdegoes commented on src/Controller/Driver.purs:
```diff
@@ -0,0 +1,227 @@
+instance routeModifierSortSetter :: RS.RouteModifier SortSetter where
+  toHashModifier (SortSetter sort) =
+    Rgx.replace sortRgx res
+    where res = "$1" <> M.sort2string sort 
```
Oh, we so need a type safe regex PureScript library. :smile:

jdegoes commented on src/Controller/Driver.purs:
```diff
@@ -0,0 +1,227 @@
+getPath :: String -> String
+getPath hash = 
+  fromMaybe "" do
+    matches <- Rgx.match qRgx hash
```
There's a ton of regex parsing in here, is it all related to URL / URI / path parsing? Could make sense to write an FFI function to parse it out into the traditional components of `Location`.

jdegoes replied:
Or a URI / path parser library?

jdegoes commented on src/Controller/Input.purs:
```diff
@@ -0,0 +1,56 @@
+        mkBreadcrumbs :: String -> [M.Breadcrumb]
+        mkBreadcrumbs path =
+          A.reverse $ foldl foldFn [M.rootBreadcrumb] parts
+          where parts = A.filter ((/=) "") $ Str.split "/" path
```
Again, all this string manipulation of Paths suggests we should have a `Path` abstraction which represents this information in a structured, canonical form -- I think it'd clean up and consolidate this code. Especially if we add parsers and unparsers for it.

jdegoes commented on src/Controller/Input.purs:
```diff
@@ -0,0 +1,56 @@
+        mkBreadcrumbs path =
+          A.reverse $ foldl foldFn [M.rootBreadcrumb] parts
+          where parts = A.filter ((/=) "") $ Str.split "/" path
+                foldFn (head:tail) a =
```
`head:tail` is deprecated and has very bad performance characteristics (O(n) in the size of the array). I am preferring `List` recently.

jdegoes commented on src/Controller/Request.purs:
```diff
@@ -0,0 +1,173 @@
+      RS.modifyRoute $ Cd.PathSetter b.link
+
+    -- clicked on _File_ link triggering file uploading
+    M.UploadFile node state -> do
```
For unused parameters, I prefer `_`. `state` is bound but not used here. See other places.

jdegoes commented on src/Controller/Request.purs:
```diff
@@ -0,0 +1,173 @@
+
+    -- clicked on _Folder_ link, create phantom folder
+    M.CreateFolder state -> do
+      let name = getNewName Config.newFolderName state
```
It seems like passing the whole state around may cause race conditions in the presence of concurrency. If two actions get executed at once, they'll be operating off independent states -- right?

cryogenian replied:
Right :disappointed: 

`state` is used to provide unique name for new folder. But it's definitely will cause race conditions in multiuser environment. I think options are
* don't get item list from state, get it from ajax response. There definitely will be less conflicts.
* add sync method in api. I.e. `getNewName :: String -> {name :: String}` registering new temporary name as blocked and not available for other clients. 
* change semantics of `PUT` to return new unique name after creating file. (will work only for real objects not directories)

jdegoes replied:
I'm not all that concerned about *server-side race conditions*. I'm more concerned about *client-side race conditions*, i.e. the user performing two asynchronous actions which both start from the same state and base something off it, in that case, "last one wins" and the UI may become disconnected from the (e.g.) file system or whatever the UI is controlling.

jdegoes commented on src/Utils/Event.purs:
````diff
@@ -1,6 +1,4 @@
--- | Custom event module. Probably would be moved to ```simple-dom```
--- | or other package
-module Control.Reactive.Event where
+module Utils.Event (raiseEvent) where
````
The functions here have types that lie (e.g. detail promises to return any `a`). We should find a way to make this type safe or push it into Halogen.

jdegoes commented on src/Utils/File.purs:
```diff
@@ -68,13 +47,11 @@ function append2FormData(name) {
   };
 }
 """ :: forall e a. String -> a -> FormData -> Eff e FormData 
```
These signatures don't accurately track effect types, `forall e. Eff e a` is pure (which is not the case with some of these functions). We need to tighten these signatures down so they accurately track effects. Also, as a consideration, if you like I can create `purescript-files` if you want to stuff this in a library (eventually that's the best decision, no rush now if you don't feel like doing the extraction).

jdegoes commented:
OK, I've left a bunch of comments, after these are resolved, addressed, or deferred to another PR, I'm happy to merge, although I'd like a review by @garyb or @puffnfresh first! Thanks Maxim.

garyb commented:
From a "cold read" I don't have any comments in addition to what has already been brought up. I'm going get up to speed with where everything is at over the next few days so should be able to offer more useful feedback in the future.

puffnfresh commented:
The translation to Halogen is awesome :+1: 

It looks like it also has additional functionality, e.g. we can now press the Folder and Notebook buttons and have additional items appear in the folder. It doesn't seem like these are hooked up to the backend. Not sure what the spec or progress is for this stuff so can't really evaluate.

cryogenian commented:
#### Fixes
* switched to affjax 
* moved api calls in `Driver` to separate `Aff` 
* default names `Untitled Notebook.slam` and `Untitled Folder`
* unused variables -> `_`
* managing double clicks in item toolbar. It raises an error to console `Failed pattern match`. Solution depends on slamdata/purescript-halogen#59
* removed unsafe/unneeded funcs (like `details` in `Utils.Event`) 
* switched to `Aff` in `Utils.File` 
* safer encoding/decoding

#### About state and client race conditions
Before sending ajax request with file content, raised `forkAff (pure $ M.ItemAdd item)` where item is phantom. After success or error from main thread phantom item deleted. Name conflicts are resolved because client uses `state.items` for name constructing, and there is new item with such name added _almost_ syncronously.

Other options: 
* generate names on server. 
* use guid as suffix for new items.

jdegoes commented:
The race conditions can be avoided with one simple rule: simply do not return state from an event. That is, treat the state as "read only". You can generate events, but the events should not contain new state which replaces the old state blindly. Instead, the events should contain "information", which the view uses to update the state in an intelligent manner. In rare occasions, if you need to update the state directly from an event, you should use emit state function `s -> s`, which avoids all race issues. 

Does that make sense, and seem like a good guideline?

Finally, @puffnfresh or @garyb, have any comments before I merge this in???

garyb commented:
> The race conditions can be avoided with one simple rule: simply do not return state from an event. That is, treat the state as "read only". You can generate events, but the events should not contain new state which replaces the old state blindly.

I strongly agree with this, I've been bitten by races in state-writing event updates quite a few times on a recent project, they can be very tricky to track down.

puffnfresh commented:
:shipit:
